### PR TITLE
Improve module augmentation in `table` plugin

### PR DESCRIPTION
**Pull request by filipsobol**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal: Improve TypeScript configuration in the project and module augmentation in `table` plugin. Related to #13565.

### CODE REVIEW

File: packages/ckeditor5-table/tsconfig.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
-  "extends": "ckeditor5/tsconfig.json",
+  "extends": "../../tsconfig.json",
```
Comment (niegowski) [inline packages/ckeditor5-table/tsconfig.json:2]: Why such change?

File: packages/ckeditor5-table/plugin.d.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import type {
```
Comment (niegowski) [inline packages/ckeditor5-table/plugin.d.ts:6]: Why this is `d.ts`? Maybe this could be a standard `ts` file?

**Review by niegowski — CHANGES_REQUESTED** [review]

**Comment by niegowski** [comment]
I'm also concerned about automatic tests for other packages. Those import other features to test integrations and they probably do not import using `index` files. Those could fail if augmentations are not provided.

File: packages/ckeditor5-table/plugin.d.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		[ PlainTableOutput.pluginName ]: PlainTableOutput;
+	}
+
+  interface CommandsMap {
+		insertTableColumnLeft: InsertColumnCommand;
+		insertTableColumnRight: InsertColumnCommand;
+    insertTableRowAbove: InsertRowCommand;
+		insertTableRowBelow: InsertRowCommand;
+    insertTable: InsertTableCommand;
+    mergeTableCellRight: MergeCellCommand;
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-table/plugin.d.ts]: Fixed

File: packages/ckeditor5-table/plugin.d.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    [ TableMouse.pluginName ]: TableMouse;
+    [ TableProperties.pluginName ]: TableProperties;
+		[ TablePropertiesEditing.pluginName ]: TablePropertiesEditing;
+		[ TablePropertiesUI.pluginName ]: TablePropertiesUI;
+    [ TableSelection.pluginName ]: TableSelection;
+    [ TableToolbar.pluginName ]: TableToolbar;
+    [ TableUI.pluginName ]: TableUI;
+    [ TableUtils.pluginName ]: TableUtils;
+		[ PlainTableOutput.pluginName ]: PlainTableOutput;
+	}
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-table/plugin.d.ts]: Fixed

File: packages/ckeditor5-table/src/index.ts [hunk 5]
```diff
@@ -29,3 +29,36 @@ export { default as TableSelection } from './tableselection';
 export { default as TableUtils } from './tableutils';
 export { default as TableColumnResize } from './tablecolumnresize';
 export { default as TableColumnResizeEditing } from './tablecolumnresize/tablecolumnresizeediting';
+
+export type { TableConfig } from './tableconfig';
+export type { default as InsertColumnCommand } from './commands/insertcolumncommand';
```
Comment (Reinmar) [inline packages/ckeditor5-table/src/index.ts:34]: Why only export types? AFAIK we usually export entire commands. They are a public API of a feature (both, as registered in commands registry and classes that can be reused).

File: packages/ckeditor5-table/src/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import type {
```
Comment (Reinmar) [inline packages/ckeditor5-table/src/plugin.ts:6]: This looks like a duplication of index.ts. Do we really need that?

File: packages/ckeditor5-table/src/plugin.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
+/**
```
Comment (Reinmar) [inline packages/ckeditor5-table/src/plugin.ts:1]: As I commented in the PR thread, I'd consider renaming this to `package.ts` as it contains kind of metadata for a package. But then I'd merge the `fooconfig.ts` file into this one too. Why having configs in one place and plugin and command maps in another?

**Comment by Reinmar** [comment]
My biggest doubt regarding this is: why do we have `plugin.d.ts` and `index.ts`? Why not merging them into one file? And keeping there all the module augmentation too?

File: packages/ckeditor5-table/tsconfig.json (hunk 1 above)
Comment (filipsobol) [reply packages/ckeditor5-table/tsconfig.json:2]: To make it consistent with `tsconfig.release.json` in the same package.

File: packages/ckeditor5-table/src/plugin.ts (hunk 6 above)
Comment (filipsobol) [reply packages/ckeditor5-table/src/plugin.ts]: On today's daily, the team agreed to keep module augmentation in a separate `augmentation.ts` file.

File: packages/ckeditor5-table/src/plugin.ts (hunk 7 above)
Comment (filipsobol) [reply packages/ckeditor5-table/src/plugin.ts]: On today's daily, the team agreed to use the `augmentation.ts` file name and to store config in a separate file.

File: packages/ckeditor5-table/src/index.ts (hunk 5 above)
Comment (Reinmar) [reply packages/ckeditor5-table/src/index.ts]: F2F discussion: It turned out that we didn't expose any commands so far in index.js files. I thought we did expose some already. Hence, let's continue not exposing anything for now (except types if needed).

**Comment by Reinmar** [comment]
Note: Squash all commits in this branch before the merge so to not pollute the history with all the test import changes.

**Comment by filipsobol** [comment]
Adding `"./packages/*/src/augmentation.ts"` to `include` array in `tsconfig.json` didn't work when running `testautomated.js`. However, it worked when we added `"./packages/ckeditor5-table/src/augmentation.ts"` to `files`.

When debugging this with @niegowski and comparing behavior of `tsc` and `testautomated.js`, we found that `tsc` worked with both configurations. After we enabled the `logLevel: 'info'` option in `ts-loader` we noticed that it loads `tsconfig.js` for `paragraph` plugin, despite testing `clipboard` plugin in root of the monorepo project. When we added `"./packages/*/src/augmentation.ts"` to `include` array in `./packages/ckeditor5-paragraph/tsconfig.json`, it started working as expected for `clipboard`.

We need to investigate why seemingly random `tsconfig.json` is being loaded by `ts-loader`. For now, as a quick workaround, we will add `"./packages/*/src/augmentation.ts"`to `include` arrays in all packages.

**Comment by filipsobol** [comment]
Above issue with the wrong \`tsconfig.json\` being loaded will be fixed in ckeditor/ckeditor5-dev#847

**Comment by Reinmar** [comment]
Note to self: I wanted to comment that you changed formatting of JSON files and I remember that we have a special rule for them in `.editorconfig` to use two spaces. But then I checked it's for `package.json`s only :D And we actually use tabs in other JSON files. Case closed :D

**Review by niegowski — APPROVED** [review]

**Comment by niegowski** [comment]
Please update the "Suggested merge commit message" to address changes in this PR.

File: tsconfig.test.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+	"extends": "./tsconfig.json",
+	"include": [
+		"./typings/",
+		"./packages/*/src/augmentation.ts",
+		"./external/*/packages/*/src/augmentation.ts"
```
Comment (pomek) [inline tsconfig.test.json:6]:
```suggestion
		"./external/ckeditor5-internal/packages/*/src/**/*.ts",
		"./external/collaboration-features/packages/*/src/**/*.ts",
```

To match https://github.com/ckeditor/ckeditor5/blob/0e5ddd74e5684897d2d2c87a0e65150c287433d4/tsconfig.docs.json#L5-L6. People can have more than two repositories in `external/` and it should. not be loaded.

Btw, we use spaces in all JSON files.

Comment (filipsobol) [reply tsconfig.test.json]: I updated it to:  
 

```
"./external/ckeditor5-internal/packages/*/src/augmentation.ts",
"./external/collaboration-features/packages/*/src/augmentation.ts"
```

File: scripts/check-manual-tests.sh [hunk 9]
```diff
@@ -26,6 +27,11 @@ then
   MANUAL_TEST_SERVER_OPTIONS="$MANUAL_TEST_SERVER_OPTIONS -i $IDENTITY_FILE"
 fi
 
+if [ ! -z "$TS_CONFIG" ]
+then
+  MANUAL_TEST_SERVER_OPTIONS="$MANUAL_TEST_SERVER_OPTIONS --tsconfig $TS_CONFIG"
+fi
```
Comment (pomek) [inline scripts/check-manual-tests.sh:33]: I would add `--tsconfig ./tsconfig.test.json` to L43 and remote support for the `-t` flag. This script is used only for verification on CI and when releasing stuff. We do not have customize it.

**Review by pomek — APPROVED** [review]
LGTM. But TBH, too many changes. I will depend on CI results.

**Pull request merged**